File: internal/nano/task_induction.go
package nano

// task_induction.go is the canonical induction-head training setup: a random block repeated twice
// (r ++ r) with DENSE next-token supervision at every position. In the second copy, the next token
// is predictable by induction (copy the token that followed the same token in the first copy).
//
// The key lesson this encodes: induction needs a *previous-token head*, which only forms under DENSE
// next-token supervision — not the single-answer-token supervision used by the assoc/retrieval tasks.
// 2-layer transformers solve this reliably; if our engine does too, the recall circuit works and the
// retrieval plateau was a supervision-density problem.

// InductionTask: vocabulary V, block length m (sequence length L = 2m).
type InductionTask struct{ V, M, L int }

// NewInductionTask builds a repeated-block induction task with block length m.
func NewInductionTask(v, m int) InductionTask { return InductionTask{V: v, M: m, L: 2 * m} }

func (t InductionTask) sample(rng *rng) []int {
	seq := make([]int, t.L)
	for i := 0; i < t.M; i++ {
		tok := int(rng.next() % uint64(t.V))
		seq[i] = tok
		seq[i+t.M] = tok
	}
	return seq
}

// Batch builds B sequences with DENSE next-token targets (target[i] = seq[i+1]).
func (t InductionTask) Batch(B int, seed int64) (idx, targets []int) {
	rng := newRNG(seed)
	idx = make([]int, B*t.L)
	targets = make([]int, B*t.L)
	for b := 0; b < B; b++ {
		seq := t.sample(rng)
		copy(idx[b*t.L:], seq)
		for i := 0; i < t.L; i++ {
			if i < t.L-1 {
				targets[b*t.L+i] = seq[i+1]
			} else {
				targets[b*t.L+i] = -1
			}
		}
	}
	return
}

// Accuracy measures next-token accuracy over the SECOND copy (induction-predictable positions),
// where seq[i] first occurred at i-M and was followed by seq[i-M+1] == seq[i+1].
func (t InductionTask) Accuracy(g *GPT, seed, n int) float64 {
	rng := newRNG(int64(seed))
	correct, total := 0, 0
	for s := 0; s < n; s++ {
		seq := t.sample(rng)
		for i := t.M; i < t.L-1; i++ { // second copy, excluding the very last token
			total++
			if g.PredictAt(seq, 1, t.L, i) == seq[i+1] {
				correct++
			}
		}
	}
	return float64(correct) / float64(total)
}

File: src/nano/task_induction.rs
//! The canonical induction-head training setup: a random block repeated twice (r ++ r) with DENSE
//! next-token supervision at every position. Induction needs a previous-token head, which only
//! forms under dense supervision — the lesson behind the retrieval plateau fix.

use super::model::{Gpt, Rng};

/// InductionTask: vocabulary V, block length M (sequence length L = 2M).
#[derive(Clone, Copy)]
pub struct InductionTask {
    pub v: usize,
    pub m: usize,
    pub l: usize,
}

impl InductionTask {
    /// NewInductionTask builds a repeated-block induction task with block length m.
    pub fn new(v: usize, m: usize) -> InductionTask {
        InductionTask { v, m, l: 2 * m }
    }

    fn sample(&self, rng: &mut Rng) -> Vec<usize> {
        let mut seq = vec![0usize; self.l];
        for i in 0..self.m {
            let tok = (rng.next() % self.v as u64) as usize;
            seq[i] = tok;
            seq[i + self.m] = tok;
        }
        seq
    }

    /// Batch builds B sequences with DENSE next-token targets (target[i] = seq[i+1]).
    pub fn batch(&self, b: usize, seed: i64) -> (Vec<usize>, Vec<i32>) {
        let mut rng = Rng::new(seed);
        let mut idx = vec![0usize; b * self.l];
        let mut targets = vec![0i32; b * self.l];
        for bb in 0..b {
            let seq = self.sample(&mut rng);
            idx[bb * self.l..bb * self.l + self.l].copy_from_slice(&seq);
            for i in 0..self.l {
                if i < self.l - 1 {
                    targets[bb * self.l + i] = seq[i + 1] as i32;
                } else {
                    targets[bb * self.l + i] = -1;
                }
            }
        }
        (idx, targets)
    }

    /// Accuracy measures next-token accuracy over the SECOND copy (induction-predictable positions).
    pub fn accuracy(&self, g: &Gpt, seed: i64, n: usize) -> f64 {
        let mut rng = Rng::new(seed);
        let mut correct = 0;
        let mut total = 0;
        for _ in 0..n {
            let seq = self.sample(&mut rng);
            for i in self.m..self.l - 1 {
                total += 1;
                if g.predict_at(&seq, 1, self.l, i) == seq[i + 1] {
                    correct += 1;
                }
            }
        }
        correct as f64 / total as f64
    }
}
